demos/coreHTTP/http_demo_s3_download_multithreaded.c

```diff
@@ -0,0 +1,965 @@
+                                           QueueHandle_t xRequestQueue,
+                                           QueueHandle_t xResponseQueue )
+{
+    BaseType_t xStatus = pdPASS;
+    size_t xRequestCount = 0;
+
+    /* Configurations of the initial request headers. */
+    HTTPRequestInfo_t xRequestInfo = { 0 };
+
+    /* The length of the file at democonfigS3_PRESIGNED_GET_URL. */
+    size_t xFileSize = 0;
+
+    /* The number of bytes we want to request within each range of the file. */
+    size_t xNumReqBytes = 0;
+    /* xCurByte indicates which starting byte we want to download next. */
+    size_t xCurByte = 0;
+
+    /* Initialize the request object. */
+    xRequestInfo.pHost = pcHost;
+    xRequestInfo.hostLen = xHostLen;
+    xRequestInfo.pMethod = HTTP_METHOD_GET;
+    xRequestInfo.methodLen = httpexampleHTTP_METHOD_GET_LENGTH;
+    xRequestInfo.pPath = pcRequest;
+    xRequestInfo.pathLen = xRequestUriLen;
+
+    /* Set "Connection" HTTP header to "keep-alive" so that multiple requests
+     * can be sent over the same established TCP connection. This is done in
+     * order to download the file in parts. */
+    xRequestInfo.reqFlags = HTTP_REQUEST_KEEP_ALIVE_FLAG;
+
+    /* Get the length of the S3 file. */
+    xStatus = prvGetS3ObjectFileSize( &xRequestInfo,
+                                      xRequestQueue,
+                                      xResponseQueue,
+                                      &xFileSize );
+
+    if( xFileSize < democonfigRANGE_REQUEST_LENGTH )
+    {
+        xNumReqBytes = xFileSize;
+    }
+    else
+    {
+        xNumReqBytes = democonfigRANGE_REQUEST_LENGTH;
+    }
+
+    /* Here we iterate sending byte range requests and retrieving responses
+     * until the full file has been downloaded. We keep track of the next byte
+     * to download with xCurByte. When this reaches the xFileSize we stop
+     * downloading. We keep track of the number of responses we are waiting for
+     * with xRequestCount.
+     */
+    while( ( xStatus != pdFAIL ) && ( xCurByte < xFileSize || xRequestCount > 0 ) )
+    {
+        /* Send range request if remaining. */
+        if( xCurByte < xFileSize )
+        {
+            xStatus = prvRequestS3ObjectRange( &xRequestInfo,
+                                               xRequestQueue,
+                                               xCurByte,
+                                               xCurByte + xNumReqBytes - 1 );
+
+            if( xStatus == pdPASS )
+            {
+                xRequestCount += 1;
+                xCurByte += xNumReqBytes;
+
+                if( ( xFileSize - xCurByte ) < xNumReqBytes )
+                {
+                    xNumReqBytes = xFileSize - xCurByte;
+                }
+            }
+        }
+
+        /* Retrieve response. */
+        if( ( xRequestCount > 0 ) && ( xStatus != pdFAIL ) )
+        {
+            xStatus = prvRetrieveHTTPResponse( xResponseQueue, &xResponseItem );
```

Comment:
What happens if we hit a timeout waiting for a response? Do we consider that as a failure?

Reply:
The response is received in `HTTPClient_Send()` itself, so an error relating to a missing response would occur in the HTTP thread, before this line (and is treated as a failure).

Reply:
I see. Thanks. Is this waiting in the main task until a response arrives from the HTTP task was an explicit design choice?
I see that main task waits for a successful response for a request in a synchronized way.
One other choice would have been to not wait for the response, but just check for it and keep posting to the HTTP task until the queue has some space.
It is just a question. No action is required if this was the choice made.

Reply:
Thanks for your comment -- after looking at it in more depth, I found a nuance in the design of the original demo that I hadn't translated accurately in the port. Mainly, I have made a change in the while loop to not fail the download on the initial failure to receive a response, but rather just check for one after each request, and made a couple of follow-up changes as a result. Let me know what you think of the modified implementation -- thanks.